### PR TITLE
Add semicolon preference to formatter options

### LINKED ISSUE #32831: Improve semicolon detection experience for new files
**TypeScript Version:**  3.6.0-dev.20190810

**Search Terms:**
- semicolon
- format
- auto import

**Repo**

For a very simple project:

`tsconfig.json`
```json
{
    "compilerOptions": {
        "module": "commonjs",
        "target": "es2016",
        "jsx": "preserve"
    },
    "exclude": [
        "node_modules",
    ]
}
```

`constants.ts`
```ts
export const a = 1
export const b = 1
export const c = 1
export const d = 1
export const e = 1
export const f = 1
```

`main.ts`
```ts
console.log(a)
```

1. Trigger auto import quick fix in `main.ts` on `a`

**Expected behavior:**
Since `constants.js` does not use semicolons, the import should not have a semi colon

**Actual behavior:**
The added import has a semi colon

I believe this is because the current semicolon detection logic (#31801) requires some number of examples in the current file to learn from. For new files, this means we always default to using semicolons even if the rest of the project is semicolonless

I can understand why we do this, but as a user I found it frustrating that TS would still insert semicolons until I added a certain number of imports (or other statements that are also used by the detection code). I don't know if looking at the rest of your project code is feasible or not for semicolon detection but we should discuss if we can improve the experience for new files

/cc @andrewbranch 

**Related Issues:** 
- #19882

### PR DESCRIPTION
**Pull request by andrewbranch**
Fixes #32831 
Related: #18780 

Corresponding VS Code PR: https://github.com/microsoft/vscode/pull/80828

Adds a first-class formatter option for optional semicolons, with three options:

1. **Ignore** (default): Don’t touch semicolons or lackthereof.
2. **Insert**: Add a real semicolon wherever ASI would.
3. **Remove**: Remove semicolons that are optional due to ASI.

Because new and reprinted nodes must come out of the writer with semicolons by default, a setting of “Ignore” will try to pick whether to keep or remove semicolons based on the current file content. (This is the same behavior as currently exists via #31801/#32903, but uses the formatter instead of a wrapper around EmitTextWriter. The change here is that the “smart” behavior can be _disabled_ by picking a preference in the formatter.)

I tried this out on various files in the TypeScript codebase including checker.ts, validating by swapping our ESLint semicolon rule from `always` to `never`. The formatter successfully removed all semicolons and the file was lint-free, and same for doing the reverse.

Screen capture of running semicolon removal then insertion on a large source file

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -9,5 +9,7 @@
         "rulePaths": ["../scripts/eslint/built/rules/"],
         "ext": [".ts"]
     },
-    "eslint.workingDirectories": ["./src", "./scripts"]
+    "eslint.workingDirectories": ["./src", "./scripts"],
+    "typescript.tsserver.log": "normal",
```
Comment (weswigham) [inline .vscode/settings.json]: ???

File: .vscode/settings.json [hunk 2]
```diff
@@ -9,5 +9,7 @@
...
+    "typescript.tsdk": "built/local"
```
Comment (weswigham) [inline .vscode/settings.json]: ???

File: .vscode/settings.json (hunk 1 above)
Comment (andrewbranch) [reply .vscode/settings.json]: Gah sorryyyy

File: .vscode/settings.json (hunk 2 above)
Comment (andrewbranch) [reply .vscode/settings.json]: 🙈

File: src/services/formatting/formatting.ts [hunk 3]
```diff
@@ -1120,7 +1148,11 @@ namespace ts.formatting {
 
         function recordDelete(start: number, len: number) {
             if (len) {
-                edits.push(createTextChangeFromStartLength(start, len, ""));
+                const newEdit = createTextChangeFromStartLength(start, len, "");
+                const lastEdit = lastOrUndefined(edits);
+                if (!lastEdit || !textSpansEqual(lastEdit.span, newEdit.span) || lastEdit.newText !== newEdit.newText) {
```
Comment (andrewbranch) [inline src/services/formatting/formatting.ts]: I don't like this; I think I have an idea to avoid it

File: .vscode/settings.json (hunk 1 above)
Comment (weswigham) [reply .vscode/settings.json]: It's alright, this bad boy only got committed to the repo with the eslint migration, and this problem is _exactly_ why we didn't commit it before - there's no concept of "per-user workspace settings", so this kinda falls flat. @mjbvz https://github.com/microsoft/vscode/issues/1435 or https://github.com/microsoft/vscode/issues/15909 when - being able to have a `settings.json` in the `.vscode` folder and a `settings.user.json` alongside it would be very nice.

Comment (andrewbranch) [reply .vscode/settings.json]: Ok, I thought it had been ignored before. I always keep my tsdk the local one so I can test my changes immediately.

File: src/compiler/core.ts [hunk 4]
```diff
@@ -329,6 +329,21 @@ namespace ts {
         return undefined;
     }
 
+    /**
+     * Like `forEach`, but iterates in reverse order.
+     */
+    export function forEachRight<T, U>(array: readonly T[] | undefined, callback: (element: T, index: number) => U | undefined): U | undefined {
```
Comment (amcasey) [inline src/compiler/core.ts]: This is probably just a convention I don't recognize, but why "right", rather than "reverse"?

File: src/compiler/core.ts [hunk 5]
```diff
@@ -2159,8 +2174,19 @@ namespace ts {
         return (arg: T) => f(arg) && g(arg);
     }
 
-    export function or<T>(f: (arg: T) => boolean, g: (arg: T) => boolean): (arg: T) => boolean {
-        return arg => f(arg) || g(arg);
+    export function or<T extends unknown>(...fs: ((arg: T) => boolean)[]): (arg: T) => boolean {
+        return arg => {
+            for (const f of fs) {
```
Comment (amcasey) [inline src/compiler/core.ts]: How do you choose between for-of, for-increment, and `forEach`?

File: src/server/protocol.ts [hunk 6]
```diff
@@ -2956,6 +2956,12 @@ namespace ts.server.protocol {
         Smart = "Smart",
     }
 
+    export enum SemicolonPreference {
```
Comment (amcasey) [inline src/server/protocol.ts]: Some of the other formatting settings have an auto value that attempts to infer an appropriate behavior from other text in the file. Would that make sense for semicolons?

File: src/server/protocol.ts [hunk 7]
```diff
@@ -2956,6 +2956,12 @@ namespace ts.server.protocol {
         Smart = "Smart",
     }
 
+    export enum SemicolonPreference {
+        Ignore = "ignore",
```
Comment (amcasey) [inline src/server/protocol.ts]: I assume we have to use ignore for this repo, but what does the diff look like if you choose insert?

File: src/compiler/core.ts (hunk 4 above)
Comment (andrewbranch) [reply src/compiler/core.ts]: Yep, I was modeling off of `Array.prototype.reduceRight`. Without a precedent of convention I would avoid using “handed” language as I imagine it’s somewhat dissonant to folks whose first written language is RTL.

File: src/compiler/core.ts (hunk 5 above)
Comment (andrewbranch) [reply src/compiler/core.ts]: My personal tendencies:

- for-of: lets you early return enclosing function, doesn’t give you element index
- for-increment: generally only use when element index is important, or if I need to start somewhere other than the first element
- `forEach`: gives you early termination of iteration but not return of the enclosing function, result transformation, and element index. It’s generally replaceable by for-of but is more elegant if you return something, like

```ts
const firstMatchingSignature = forEach(types, type => {
  return firstDefined(type.getCallSignatures(), somePredicateThing);
});
```

- `Array.prototype.forEach`: no early termination, no return value 👎

File: src/server/protocol.ts (hunk 6 above)
Comment (andrewbranch) [reply src/server/protocol.ts]: Some of the _user preferences_ have “auto” settings, but none of the formatter settings do, as far as I’ve seen. The distinction is pretty confusing and my guess is the user preferences that deal with formatting arose from the limitations of the formatter. (Before this PR, the formatter dealt _exclusively_ with whitespace.) The distinction isn’t _really_ surfaced to users, so it may not matter much, but it would be nice if we had a more unified story around where this kind of functionality lives. As such, I thought about it, but I’m not sure a _formatter_ is the correct place for inference. That said, when writing new nodes (like auto imports), if you’ve chosen “ignore” (the default), we do make an attempt to match your current file contents. The idea is that when we generate new nodes, we _have_ to make a decision somehow: either the text is going to have semicolons or it’s not.

There’s theoretically nothing stopping us from introducing an “auto” setting for the formatter too, but I’d rather wait and see if people ask for it, purely because I wanted to avoid creating an unnecessarily confusing amount of choices.

File: src/server/protocol.ts (hunk 7 above)
Comment (andrewbranch) [reply src/server/protocol.ts]: Generally the same. I sampled some changes after auto-inserting semicolons, and I think the only pattern I saw was in single-line object types, `type X = { x: string };` becomes `type X = { x: string; };`

File: src/compiler/core.ts (hunk 5 above)
Comment (amcasey) [reply src/compiler/core.ts]: `arg => forEach(fs, f => f(arg)) || false`?

File: src/server/protocol.ts (hunk 6 above)
Comment (amcasey) [reply src/server/protocol.ts]: My only concern with the wait-and-see approach is that VS has a per-user setting for this (vs per-project) so there's no way to disable it if the current TS version doesn't support "auto".  That means that we need to have a good default on the server and it's not clear to me what it would be - ignore?

Comment (andrewbranch) [reply src/server/protocol.ts]: VS Code’s setting is also per-user. The default is “ignore,” because that’s the pre-PR behavior. “Auto” would definitely be more disruptive than “ignore.” Am I understanding your concern correctly?

**Comment by andrewbranch** [comment]
I’m starting to feel like doing this through the existing formatter/rule infrastructure was an awkwardly roundabout approach, and it would be cleaner to add some kind of “formatter preprocessors” concept where a preprocessor operates similarly to a codefix; it gets some context (selection/range/file, formatter options) and returns some TextChanges that get applied before the main formatter (which could continue to operate exclusively on whitespace) runs. (Although separate preprocessors could potentially step on each others’ toes, which could get messy. One advantage of the approach I’ve taken here is that conflicting rules trying to take effect at the same position are properly excluded.)

File: src/compiler/core.ts (hunk 5 above)
Comment (andrewbranch) [reply src/compiler/core.ts]: Yeah, that would work. I think I also have a tendency to implement abstractions with building blocks that are _less_ abstract than the function I’m implementing, if that makes sense. `or` and `forEach` are both fairly low-level atomic abstractions, so it makes sense to me that neither would depend on the other, but would rather be built with primitive loops. (I’ve never been consciously aware of this thought process; it’s typically just driven by a vague intuition.)

File: src/server/protocol.ts (hunk 6 above)
Comment (jessetrinity) [reply src/server/protocol.ts]: That sounds right to me - the default setting should be to ignore any changes to semicolon formatting.

Comment (amcasey) [reply src/server/protocol.ts]: Sorry, "default" was confusing in that context.  What would the fallback setting be if the user specified "auto" and the server didn't support it?

File: src/compiler/core.ts (hunk 5 above)
Comment (amcasey) [reply src/compiler/core.ts]: I guess, in my mind, if `or` can straightforwardly be reduced to `forEach` it is higher-level.  `forEach` also seems like the sort of thing that might eventually be performance-tuned and it would be nice to pick up the benefits automatically.  I don't feel strongly about it though.

File: src/server/protocol.ts (hunk 6 above)
Comment (andrewbranch) [reply src/server/protocol.ts]: Oh, got it. The fallback would be “ignore.” There are a number of preferences/rules that have a disclaimer about what version of TypeScript they require. It’s not the end of the world, but it’s not ideal.

It just gives me an uncomfortable squirmy feeling thinking about an option that means, “I want you to fix up all my semicolons, but I _don’t_ want to tell you how I like them!”

Of course, that has a highly practical application: “Please adapt my semicolon habits to match whatever project I happen to be working on right now.” But, that request may as well apply to _every_ formatter option, not just semicolon preference. As such, I think this is the wrong venue to try to solve the problem of “I work on a bunch of different projects with different formatting conventions.”

Comment (amcasey) [reply src/server/protocol.ts]: I agree that it would make more sense to add "auto" everywhere at once and I don't think we're ready to do that, so I'm fine with omitting "auto" for now.

**Comment by andrewbranch** [comment]
Update: this now removes the last semicolon from single-line object-like types (that are run through the formatter—doesn’t effect compiler emit):

```diff
- type X = { w: string; x: { y: { z: string; }; }; };
+ type X = { w: string; x: { y: { z: string } } };
```

(To be more specific, when the semicolon preference is “I generally like semicolons,” it removes them from synthesized nodes that get inserted and it doesn’t try to insert them into pre-existing user code on “format document.”)

**Comment by andrewbranch** [comment]
> Update: this now removes the last semicolon from single-line object-like types (that are run through the formatter—doesn’t effect compiler emit):

Update: Reverted this because I think it makes more sense to do this in the emitter. Will probably look at that in a separate PR later.

File: tests/cases/fourslash/formatAddSemicolons1.ts [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+////const x = function() { }
+////for (let i = 0; i < 1; i++) {
+////    1
+////    2
+////}
+////do { } while (false) console.log(3)
+////function f() { }
+////class C {
+////    ["one"] = {}
+////    ["two"]
```
Comment (sandersn) [inline tests/cases/fourslash/formatAddSemicolons1.ts]: 1. what happens with existing semicolons?
2. why aren't there semicolons here in the original source? are you modelling erroneous code?
3. if so, can you add

```ts
;['three'] = {}
;['four']
```

to show what happens there?

Comment (andrewbranch) [reply tests/cases/fourslash/formatAddSemicolons1.ts]: > why aren't there semicolons here in the original source? are you modelling erroneous code?

Valid, but stupidly formatted code. This parses as `["one"] = {}["two"]`, so the test ensures that we don’t insert a semicolon that would change how the input parses.

> if so, can you add

Yeah, sure. I can tell you that the formatter won’t even consider it, as a potential spot for insertion, because it processes pairs of tokens excluding trivia, so it will see a declaration that already ends in a semicolon. But it’s nice to add realistic user code as tests along with these bananas edge cases.

File: tests/cases/fourslash/formatAddSemicolons1.ts [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+}
+do { } while (false); console.log(3);
+function f() { }
+class C {
+    ["one"] = {}
+    ["two"];
+    three: string;
+    m() { }
+    ;["three"] = {}
+        ;["four"];
```
Comment (andrewbranch) [inline tests/cases/fourslash/formatAddSemicolons1.ts]: @sandersn This is your test case. The indentation is obviously not ideal, but it’s totally unrelated to anything in this PR :/

File: tests/cases/fourslash/formatAddSemicolons1.ts (hunk 8 above)
Comment (sandersn) [reply tests/cases/fourslash/formatAddSemicolons1.ts]: > erroneous code

> stupidly formatted code

You say [tʰəme<sup>ɪ</sup>ɾo<sup>ʊ</sup>], I say [tʰəmato<sup>ʊ</sup>]!

File: src/services/formatting/rules.ts [hunk 10]
```diff
@@ -780,4 +789,79 @@ namespace ts.formatting {
...
+        if (context.contextNode.kind === SyntaxKind.InterfaceDeclaration ||
+            context.contextNode.kind === SyntaxKind.TypeAliasDeclaration
+        ) {
+            // Can’t remove semicolon after `foo`; it would parse as a method declaration:
+            //
+            // interface I {
+            //   foo;
+            //   (): void
+            // }
+            return !(isPropertySignature(context.currentTokenParent)
```
Comment (sandersn) [inline src/services/formatting/rules.ts]: this function returns true to mean "can delete", right? then I expected the predicate to be 

```ts
!isPropertySignature(current) || current.type || nextTokenKind !== SyntaxKind.OpenParenToken
```

File: src/services/formatting/rules.ts [hunk 11]
```diff
@@ -780,4 +789,79 @@ namespace ts.formatting {
...
+        ) {
+            return false;
+        }
+
+        if (context.contextNode.kind === SyntaxKind.InterfaceDeclaration ||
+            context.contextNode.kind === SyntaxKind.TypeAliasDeclaration
+        ) {
+            // Can’t remove semicolon after `foo`; it would parse as a method declaration:
+            //
+            // interface I {
```
Comment (sandersn) [inline src/services/formatting/rules.ts]: why doesn't this same rule apply to property declarations with no type and no initialiser? and why only to property signatures inside interfaces and types?

File: src/services/formatting/formatting.ts [hunk 12]
```diff
@@ -938,37 +961,42 @@ namespace ts.formatting {
...
-                        // sameLineIndent flag to notify the indenter that the indentation is within the line.
-                        if (currentParent.getStart(sourceFile) === currentItem.pos) {
-                            dynamicIndentation.recomputeIndentation(/*lineAddedByFormatting*/ true);
-                        }
-                        break;
-                    default:
-                        Debug.assert(lineAction === LineAction.None);
-                }
+            if (rules) {
+                // Apply rules in reverse order so that higher when higher priority rules (which are first in the array)
```
[the remark this answers is not in the document]
Comment (andrewbranch) [reply src/services/formatting/formatting.ts]: > so that higher when higher priority rules

🤦‍♂

File: src/services/formatting/formatting.ts [hunk 13]
```diff
@@ -644,6 +652,21 @@ namespace ts.formatting {
                 consumeTokenAndAdvanceScanner(tokenInfo, node, nodeDynamicIndentation, node);
             }
 
+            if (!node.parent && formattingScanner.isOnEOF()) {
+                const token = formattingScanner.readEOFTokenRange();
+                if (token.end <= node.end && previousRange) {
+                    processPair(
```
[the remark this answers is not in the document]
Comment (andrewbranch) [reply src/services/formatting/formatting.ts]: Yeah, it’s where all the text changes actually occur. Corresponds to the pair of tokens provided in every `rule()` in `rules.ts`.

File: src/services/formatting/rulesMap.ts [hunk 14]
```diff
@@ -13,12 +13,42 @@ namespace ts.formatting {
         return rulesMapCache;
     }
 
-    export type RulesMap = (context: FormattingContext) => Rule | undefined;
+    function getRuleActionExclusion(ruleAction: RuleAction): RuleAction {
+        let mask: RuleAction = 0;
+        if (ruleAction & RuleAction.Ignore) {
+            mask |= RuleAction.TriviaAction;
```
[the remark this answers is not in the document]
Comment (andrewbranch) [reply src/services/formatting/rulesMap.ts]: Yes, if one rule says “Ignore,” what it really means is “leave this whitespace alone.” Ignore is probably a bad name now... I’ll see if I can make this make more sense, and add some comments.

File: src/services/textChanges.ts [hunk 15]
```diff
@@ -831,28 +832,40 @@ namespace ts.textChanges {
             return (options.prefix || "") + noIndent + (options.suffix || "");
         }
 
+        function getFormatCodeSettingsForWriting(context: formatting.FormatContext, sourceFile: SourceFile): FormatCodeSettings {
+            return {
+                ...context.options,
+                // If the user has no semicolon preference defined and the file doesn’t use semicolons,
+                // make the formatter remove them. Otherwise, ignore semicolons in the formatter because
+                // the writer will insert them by default.
+                semicolons: context.options.semicolons === SemicolonPreference.Remove ||
```
[the remark this answers is not in the document]
Comment (andrewbranch) [reply src/services/textChanges.ts]: `undefined` is supposed to be treated equivalently to `Ignore`, but I agree this is horrifying, so will find a way to clean it up

File: src/services/formatting/rulesMap.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (andrewbranch) [reply src/services/formatting/rulesMap.ts]: I’ve renamed a bunch of old stuff so it should make a lot more sense now. Let me know what you think.

File: src/services/formatting/rules.ts (hunk 11 above)
Comment (andrewbranch) [reply src/services/formatting/rules.ts]: Because interfaces/types can’t have property declarations, and classes/objects can’t have method signatures (or any other member that starts with an open paren, AFAICT.

Unfortunately, due to the definition of where ASI occurs, determining whether ASI occurs while parsing is really easy, but determining whether a given semicolon in an already-parsed tree _would_ have been inserted by ASI were it not explicitly included (which is what this function tries to determine) is _extremely_ wishy washy. The definition is (paraphrasing a bit) “ASI occurs just about any time the parser can’t understand the next token without inserting a semicolon first.” Which means, to truly tell whether a semicolon was necessary or not, you have to try parsing without it, and see if you get the same result. That’s why this function appears so strange, and why the chances that this misses an opportunity to delete a semicolon is _much_ higher than the chance that `isSemicolonInsertionContext` is wrong—that one essentially just enumerates the places where `parser.ts` parses an optional semicolon.

File: src/services/formatting/rules.ts (hunk 10 above)
Comment (sandersn) [reply src/services/formatting/rules.ts]: Ah-ha! I'm pretty sure I missed the surrounding parentheses here.

**Comment by thermarthae** [comment]
> > Update: this now removes the last semicolon from single-line object-like types (that are run through the formatter—doesn’t effect compiler emit):
> 
> Update: Reverted this because I think it makes more sense to do this in the emitter. Will probably look at that in a separate PR later.

@andrewbranch Are there any plans to make it work in the near future?

**Comment by andrewbranch** [comment]
@thermarthae thanks, opened #35042 to track 👍
